closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -474,12 +576,290 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+
+    /* Library names for OpenSSL 1.1.1, 1.1.0 and symbolic links:
+     * It is important to preserve the order!!!
+     *
+     * Since there is no indication of the version of a symlink,
+     * they have to be loaded first, so as to compare with other
+     * available options.
+     * Note: On macOS 11 or later, loading the general symlink causes
+     * a fatal warning and associated abort by default, so it is
+     * omitted.
+     *
+     * The rest of the libraries are listed in descending order,
+     * which allows us to do two things:
+     * - Stop if a general symlink is loaded and we then find a
+     *   specific version that is higher.
+     * - Stop immediately if a specific version is loaded, as
+     *   anything after that will be a lower version.
+     */
+    static const char * const libNames[] = {
+#if defined(_AIX)
+        "libcrypto.a(libcrypto64.so)",      /* general symlink library name from archive file */
+        "libcrypto64.so",                   /* general symlink library name */
+        "libcrypto.a(libcrypto.so)",        /* general symlink library name from archive file */
+        "libcrypto.so",                     /* general symlink library name */
+        "libcrypto.a(libcrypto64.so.3)",    /* 3.x library name from archive file */
+        "libcrypto64.so.3",                 /* 3.x library name */
+        "libcrypto.a(libcrypto.so.3)",      /* 3.x library name from archive file */
+        "libcrypto.so.3",                   /* 3.x library name */
+        "libcrypto.a(libcrypto64.so.1.1)",  /* 1.1.x library name from archive file */
+        "libcrypto.so.1.1",                 /* 1.1.x library name */
+        "libcrypto.so.1.0.0",               /* 1.0.x library name */
+#elif defined(__APPLE__) /* defined(_AIX) */
+        "libcrypto.3.dylib",                /* 3.x library name */
+        "libcrypto.1.1.dylib",              /* 1.1.x library name */
+        "libcrypto.1.0.0.dylib",            /* 1.0.x library name */
+#elif defined(_WIN32) /* defined(__APPLE__) */
+        "libcrypto-3-x64.dll",              /* 3.x library name */
+        "libcrypto-1_1-x64.dll",            /* 1.1.x library name */
+        "libeay32.dll",                     /* old library name */
+#else /* defined(_WIN32) */
+        "libcrypto.so",                     /* general symlink library name */
+        "libcrypto.so.3",                   /* 3.x library name */
+        "libcrypto.so.1.1",                 /* 1.1.x library name */
+        "libcrypto.so.1.0.0",               /* 1.0.x library name */
+        "libcrypto.so.10",                  /* old library name */
+#endif /* defined(_AIX) */
+    };
+
+    numOfLibs = sizeof(libNames) / sizeof(libNames[0]);
+    /* If JAVA_HOME is not null or empty and no library has been loaded yet, try there. */
+    if ((NULL != chomepath) && ('\0' != *chomepath) && (NULL == crypto_library)) {
+        char **libNamesWithPath = malloc(size * sizeof(char *));
```

Comment:
Some problems remain:
* I don't see a definition of `size` which means this won't compile
* only one entry in `libNamesWithPath` is used at a time, so there's no need for an array
* `libNamesWithPath` and it's entries are not freed, nor is `libPath`
* the hard-coded value `16` should be replaced by `sizeof(suffix)` where `suffix` is suitably defined:
```c
#if defined(_WIN32)
    static const char pathSuffix[] = "\\bin\\";
#else /* defined(_WIN32) */
    static const char pathSuffix[] = "/lib/";
#endif /* defined(_WIN32) */
```
* the static locals can be declared first, allowing non-static locals to be initialized directly
```c
    const size_t numOfLibs = sizeof(libNames) / sizeof(libNames[0]);
#if defined(_AIX)
    const size_t num_of_generic = 4;
#elif defined(__linux__) /* defined(_AIX) */
    const size_t num_of_generic = 1;
#else /* defined(__linux__) */
    const size_t num_of_generic = 0;
#endif /* defined(_AIX) */

    void *result = NULL;
    void *prevResult = NULL;
    size_t i = 0;
    long tempVersion = 0;
    long previousVersion = 0;
```
* this could do a single allocation for `libPath` that has sufficient room for `chomepath`, the subdirectory, and the longest library name, replacing the library name for each iteration (but that might be harder to understand)

Reply:
Regarding the problems, here is what I did:
- I renamed `size` to `numOfLibs`, but I missed that spot.
- I avoid creating an array. I just use a single `libNameWithPath` variable in the loop.
- I added `free` commands for both where appropriate.
- I substituted the hardcoded value with appropriately calculated sizes.
- I moved declarations around as suggested to have the non-static ones be initialized directly.
- I think that would add complexity that wouldn't really get us much since this is only run in the beginning and the loop stops as soon as a library is found.